src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs

```diff
@@ -75,6 +78,36 @@ internal QuicListener CreateQuicListener(IPEndPoint endpoint)
             return CreateQuicListener(options);
         }
 
+        internal async Task PingPong(QuicConnection client, QuicConnection server)
+        {
+            using QuicStream clientStream = client.OpenBidirectionalStream();
+            ValueTask t = clientStream.WriteAsync(s_ping);
+            using QuicStream serverStream = await server.AcceptStreamAsync();
+
+            byte[] buffer = new byte[s_ping.Length];
+            int remains = s_ping.Length;
+            while (remains > 0)
+            {
+                int readLength = await serverStream.ReadAsync(buffer, buffer.Length - remains, remains);
+                Assert.True(readLength > 0);
+                remains -= readLength;
+            }
+            Assert.Equal(s_ping, buffer);
+            await t;
+
+            t = serverStream.WriteAsync(s_pong);
+            remains = s_pong.Length;
+            while (remains > 0)
+            {
+                int readLength = await clientStream.ReadAsync(buffer, buffer.Length - remains, remains);
+                Assert.True(readLength > 0);
+                remains -= readLength;
+            }
+
+            Assert.Equal(s_pong, buffer);
+            await t;
+        }
```

Comment:
We need to send FIN flag (or Shutdown) from both sides when we are done writing, otherwise Stream's Dispose wouldn't work correctly (especially after @scalablecory 's changes in https://github.com/dotnet/runtime/pull/52929)

Reply:
I don't think it should be in PingPong - as one can do more than that or it may be used multiple times. 

Could you please elaborate more on the Dispose? For purpose of the test I really only care about fact that server received correct certificate from client and that the connection/stream is useable afterwards. But I'm curious what badness may happen.

Reply:
The method now creates its own streams and disposes them after finishing -- so if the method is called again, new streams would be created. It would be another story if the streams were passed to the method from the outside, or returned from the method for someone to use, but they now clearly live only through this method.

If a stream is closed in msquic without either shutting down gracefully or aborting -- @ManickaP did investigation -- I believe this stream would still hang around inside msquic and will not be decremented from active streams count on the connection.

With Cory's changes, at least at their current state, Dispose would wait forever for SHUTDOWN_COMPLETE event, which will never arrive since streams were not shut down.

We might not care about this in tests, but I would still love if the tests contained best practices for usage... I mean we are still writing "using" in tests even though we can also say most of the time we don't care about proper disposal in tests?

Reply:
> If a stream is closed in msquic without either shutting down gracefully or aborting -- @ManickaP Máña Píchová FTE did investigation -- I believe this stream would still hang around inside msquic and will not be decremented from active streams count on the connection.

Yep, it's been discussed here: https://github.com/microsoft/msquic/discussions/1510

Reply:
@CarnaViire  that since the stream is created inside the PingPong it should be closed there. I originally envisioned PingPoing that would take two streams but then I changed it. 

I think we should fix Dispose in this case and then do what ever is needed there. Relying on users to always call right sequence is going to be problematic IMHO. It also concerns me that we cannot cleanup without peer's cooperation. 
We had issues in the past when shutdown/Dispose hang when transport get to bad state. Perhaps`QuicStreamClose ` can get some 'force' flag @nibanks or we would use QUIC_STREAM_SHUTDOWN_FLAGS.IMMEDIATE.

Reply:
The streams should be disposed with using. The shutdown behavior should be fixed with #54798.